tests/unit/test_group_project.py

```diff
@@ -0,0 +1,206 @@
+import ddt
+from unittest import TestCase
+import mock
+from group_project_v2.group_project import GroupActivityXBlock
+from xblock.runtime import Runtime
+
+
+from group_project_v2.stage_components import GroupProjectReviewQuestionXBlock
+from xblock.field_data import DictFieldData
+from tests.utils import TestWithPatchesMixin
+
+
+def _make_question(question_id):
+    question_mock = mock.create_autospec(spec=GroupProjectReviewQuestionXBlock)
+    question_mock.question_id = question_id
+    return question_mock
+
+
+def _make_reviews(reviews):
+    result = []
+    for review in reviews:
+        reviewer, question, answer = review
+        result.append({'reviewer': reviewer, 'question': question, 'answer': answer})
+    return result
+
+
+def _make_workgroup(user_ids):
+    return {'users': [{'id': user_id} for user_id in user_ids]}
+
+
+@ddt.ddt
+class TestCalculateGradeGroupActivityXBlock(TestWithPatchesMixin, TestCase):
+    def setUp(self):
+        self.project_api_mock = mock.Mock()
+        self.runtime_mock = mock.Mock(spec=Runtime)
+        self.grade_questions_mock = mock.PropertyMock()
+        self.block = GroupActivityXBlock(self.runtime_mock, field_data=DictFieldData({}), scope_ids=mock.Mock())
+
+        self.project_api_mock = self.make_patch(self.block, 'project_api')
+        self.grade_questions_mock = self.make_patch(GroupActivityXBlock, 'grade_questions', mock.PropertyMock())
+        self.real_user_id_mock = self.make_patch(self.block, 'real_user_id')
+        self.real_user_id_mock.side_effect = lambda u_id: u_id
+
+    @ddt.data(
+        (1, ["q1"], [1], [], None),
+        (2, ["q1", "q2"], [1], [], None),
+        (3, ["q1"], [1, 2], [], None),
+        (4, ["q1"], [1], [(1, "q1", 100)], 100),
+        (5, ["q1", "q2"], [1], [(1, "q1", 20), (1, "q2", 30)], 25),
+        (6, ["q1", "q2"], [1], [(1, "q1", 1), (1, "q2", 2)], round((1.0+2.0)/2.0)),  # rounding
+        (7, ["q1", "q2"], [1, 2], [(1, "q1", 1), (1, "q2", 2)], None),
+        (8, ["q1", "q2"], [1, 2], [(1, "q1", 1), (1, "q2", 2), (2, "q2", 0)], None),
+        (9, ["q1", "q2"], [1, 2], [(1, "q1", 10), (1, "q2", 20), (2, "q1", 30), (2, "q2", 60)], 30.0),
+    )
+    @ddt.unpack
+    def test_calculate_grade(self, group_id, question_ids, reviewer_ids, reviews, expected_grade):
+        self.project_api_mock.get_workgroup_reviewers = mock.Mock(
+            return_value=[{"id": rew_id} for rew_id in reviewer_ids]
+        )
```

Comment:
@smarnach I know, I just find it more explicit, and explicit is better than "ok, now I've got to go read mock documentation on how it handles missing attributes" :)